### PR TITLE
[Podspec] Add `swift_version` DSL

### PR DESCRIPTION
**Pull request by DanToml**
Relates to: https://github.com/CocoaPods/CocoaPods/issues/7134

### CODE REVIEW

File: lib/cocoapods-core/specification/root_attribute_accessors.rb [hunk 1]
```diff
@@ -40,6 +40,15 @@ def version
           end
         end
 
+        # @return [Array<Requirement>] The swift_versions required to use the
+        #                              specification.
+        #
+        def swift_versions
+          @swift_versions ||= if versions = attributes_hash['swift_versions']
```
Comment (dnkoutso) [inline lib/cocoapods-core/specification/root_attribute_accessors.rb]: do we need the `if versions = attributes_hash['swift_versions']`

Does it also need to be set in the `cocoapods_versions` method or does that have a default value always set?

File: lib/cocoapods-core/specification/dsl.rb [hunk 2]
```diff
@@ -126,6 +126,25 @@ module DSL
       root_attribute :version,
                      :required => true
 
+      #------------------#
+
+      # @!method swift_version=(version)
+      #
+      #   The version of Swift that the specification supports.
```
Comment (dnkoutso) [inline lib/cocoapods-core/specification/dsl.rb]: Could probably expand slightly more for the end user in which we document how this is used as a "best effort" value.

File: lib/cocoapods-core/specification/root_attribute_accessors.rb (hunk 1 above)
Comment (dnkoutso) [reply lib/cocoapods-core/specification/root_attribute_accessors.rb]: I think `Requirement` initializer covers the case of `nil` or empty string.

Comment (dnkoutso) [reply lib/cocoapods-core/specification/root_attribute_accessors.rb]: but i guess we would prefer a `nil` value than a `Requirement` value with `<Pod::Version version=0>`.

Comment (DanToml) [reply lib/cocoapods-core/specification/root_attribute_accessors.rb]: I prefer a nil here for the objective-c pod case

File: lib/cocoapods-core/specification/dsl.rb (hunk 2 above)
Comment (DanToml) [reply lib/cocoapods-core/specification/dsl.rb]: Yeah, planning on following up with more info and docs

**Comment by dnkoutso** [comment]
👍

**Comment by orta** [comment]
Feels like good foundations 👍

File: lib/cocoapods-core/specification/dsl.rb [hunk 3]
```diff
@@ -126,6 +126,25 @@ module DSL
...
+      #     spec.swift_version = '>= 3.2'
+      #     spec.swift_versions = ['>= 3.2', '< 4.0']
+      #
+      #   @param  [String] swift_version
+      #
+      root_attribute :swift_versions,
+                     :container => Array,
+                     :singularize => true,
+                     :multi_platform => false,
+                     :root_only => true
```
Comment (dnkoutso) [inline lib/cocoapods-core/specification/dsl.rb]: nit: I think `root_attribute` by default has this property set to true.

**Comment by siggb** [comment]
@DanToml Hello and thanks for the improvements!

Does this enhancement resolve problem when Cocoapods override `SWIFT_VERSION` in spec dependencies? (even when I use `spec.pod_target_xcconfig  = { 'SWIFT_VERSION' => '4.0' }` command `pod spec lint` fails with error `The “Swift Language Version” (SWIFT_VERSION) build setting must be set to a supported value for targets which use Swift. This setting can be set in the build settings editor.`

**Comment by siggb** [comment]
@DanToml Hello and thanks for improvements.

I've noticed a problem with spec dependencies: Cocoapods overrides `SWIFT_VERSION` on a `   Building with xcodebuild.` step.

As a result of `pod spec lint --verbose` I receive an error: `The “Swift Language Version” (SWIFT_VERSION) build setting must be set to a supported value for targets which use Swift. This setting can be set in the build settings editor.`.

**Comment by dnkoutso** [comment]
@DanToml want me to rebase/merge?

File: lib/cocoapods-core/specification/root_attribute_accessors.rb (hunk 1 above)
Comment (segiddins) [reply lib/cocoapods-core/specification/root_attribute_accessors.rb]: nil here is the right call

File: lib/cocoapods-core/specification/dsl.rb (hunk 3 above)
Comment (segiddins) [reply lib/cocoapods-core/specification/dsl.rb]: yup

File: spec/specification/root_attribute_accessors_spec.rb [hunk 4]
```diff
@@ -26,6 +26,18 @@ module Pod
...
 
+    describe '#swift_versions' do
+      it 'returns the requirements when multiple are specified' do
+        @spec.swift_versions = '>= 3.0', '< 4.0'
+        @spec.swift_versions.should == Requirement.new('>= 3.0', '< 4.0')
+      end
+
+      it 'returns the requirement when only one is specified' do
+        @spec.swift_versions = '= 3.0'
+        @spec.swift_versions.should == Requirement.new('= 3.0')
```
Comment (segiddins) [inline spec/specification/root_attribute_accessors_spec.rb]: nit: also test nil

**Comment by DanToml** [comment]
@dnkoutso waiting till I have the main pr ready too. I don’t like merging core changes till an integration pr is at least in review

**Comment by dnkoutso** [comment]
ah! ok cool sorry didn't know :)

File: lib/cocoapods-core/specification/dsl.rb (hunk 2 above)
Comment (dnkoutso) [reply lib/cocoapods-core/specification/dsl.rb]: No longer need to do this, this is just a string.

**Comment by dnkoutso** [comment]
Updated per our new design decision to scale down and not support ranges. Now `swift_version` is a simple string that is wrapped by a `Version`.

This will *always* be used if the pod author had set it, otherwise we fallback to what we do today.

PR for https://github.com/CocoaPods/CocoaPods coming up.

**Comment by dnkoutso** [comment]
Any objections from merging this?

**Comment by dnkoutso** [comment]
Going for it!
